src/test/groovy/com/netflix/frigga/conventions/sharding/ShardingNamingConventionSpec.groovy

```diff
@@ -0,0 +1,43 @@
+package com.netflix.frigga.conventions.sharding
+
+import spock.lang.Specification
+
+import static com.netflix.frigga.conventions.sharding.ShardingNamingConvention.duplicateShardWarning
+import static com.netflix.frigga.conventions.sharding.ShardingNamingConvention.shardInRemainingWarning
+
+class ShardingNamingConventionSpec extends Specification {
+
+  ShardingNamingConvention subject = new ShardingNamingConvention()
+
+  def "extracts expected shards"() {
+    given:
+    def resultOptional = Optional.ofNullable(expectedResult)
+
+    when:
+    def result = subject.extractNamingConvention(details)
+
+    then:
+    result.result == resultOptional
+    result.unprocessed == expectedRemaining
+    new ArrayList<>(result.warnings) == expectedWarnings
+
+    where:
+    details                     || expectedRemaining || expectedResult                   || expectedWarnings
+    "foo"                       || "foo"             || null                             || []
+    "x1foo-bar-baz"             || "-bar-baz"        || [1: s(1, "foo")]                 || []
+    "x1foo-x2bar-x1baz-bazinga" || "-bazinga"        || [1: s(1, "baz"), 2: s(2, "bar")] || [dupe(1, "foo", "baz")]
+    "x2bar-x1foo-blah-x3baz"    || "-blah-x3baz"     || [1: s(1, "foo"), 2: s(2, "bar")] || [remains("-blah-x3baz", 3, "baz")]
```

Comment:
Ideally we could precisely reproduce the overall resource name if we know the component parts. The support for arbitrary ordering and de-duping means that it isn't possible to do that reliably.

Reply:
as we discussed on chat - I think the fact that a NamingConvention is used to inspects a name after it has been processed into an app/stack/detail may put us in a stronger spot that we previously were (we always only use those three components (and push sequence for a deploy) to compose the final name)